RFC 474: Event Bridge Scheduler L2 (#486)
aws/aws-cdk-rfcs: text/0474-event-bridge-scheduler-l2.md

> section of your RFC. It's a powerful thought exercise which will challenge you
> to truly think about this feature from a user's point of view.
>
> Choose *one or more* of the options below:
>
> * **CHANGELOG**: Write the changelog entry for this feature in conventional
>   form (e.g. `feat(eks): cluster tags`). If this change includes a breaking
>   change, include a `BREAKING CHANGE` clause with information on how to
>   migrate. If migration is complicated, refer to a fictional GitHub issue and
>   add its contents here.
>
> * **README**: If this is a new feature, write the README section which
>   describes this new feature. It should describe the feature and walk users
>   through usage examples and description of the various options and behavior.
>
> * **PRESS RELEASE**: If this is a major feature (~6 months of work), write the
>   press release which announces this feature. The press release is a single
>   page that includes 7 paragraphs: (1) summary, (2) problem, (3) solution, (4)
>   leader quote, (5) user experience, (6) customer testimonial and (7) one
>   sentence call to action.

**CHANGELOG**:

> `feat(events-scheduler): Event Bridge Scheduler L2 constructs`

Reviewer: If possible, lets break the proposed changelog into atomic pieces. Can we introduce schedule/targets/schedule-groups separately?

Author: Could do so, but could you please explain to me what is the value of this? We will likely commit the whole implantation in a single PR to CDK repo?

Author: > The value would be that its easier to review.

What exactly easier to review, the change log?

I thought this is a proposal for how the future CDK change-log would look like. See for example L2 constructs release in https://github.com/aws/aws-cdk/releases/tag/v2.65.0. In this case, isn't it easier for the reader to have a single entry for Scheduler L2 constructs rather than multiple entries for each L2 Construct Class separately? As those construct classes on their own do not provide much value, I wouldn't break them up. We could still list the new classes in the description though.

Reviewer: I mean easier for me to review the code if the PRs are separated out into chunks. The smaller the PR, the less likely I am to miss things. If these constructs are not able to be separated then I understand, although I don't feel like that's necessarily the case.

For example, `ScheduleGroup` can and should be a separate PR imo. Just helps declutter the initial PR.

2. **Targets**: A target is an API operation that EventBridge Scheduler calls on your behalf every time your schedule runs.  Targets can include Amazon EC2 instances, AWS Lambda functions, Kinesis streams, Amazon ECS tasks, Step Functions state machines, Amazon SNS topics, Amazon SQS queues, Amazon CloudWatch LogGroups, and built-in targets. A target receives events in JSON format.
3. **Schedule Group**: Scheduling groups help you organize your schedules. Scheduling groups support tags that you can use for cost allocation, access control, and resource organization. When creating a new schedule, you can add it to a scheduling group.

This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project. It allows you to define Event Bridge Schedules.

## Defining a schedule 

```ts

> const target = new targets.LambdaInvoke(props.func, {

Reviewer: and if so, `events` already has a concept of a schedule. Can you explain what this resource achieves that the schedule class in `events` does not?

> const target = new targets.LambdaInvoke(props.func, {

Author: > this is coming from the existing events-targets module right?

No it's a separate module `targets` for schedule targets. I do not think that we should attempt to reuse the same module for EB Rules Targets and EB Schedule Targets. I think as in CFN those targets are different CFN declarations ([Rule Target CFN](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-target.html), [EB Target CFN](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-events-rule-target.html)), attempting to somewhat reuse these classes in L2 constructs will create more problems for edge-cases when behaviour between targets is different. 

Curious to know what do you think should be the reasons for reusing module targets? I would guess, that maybe in customer code it would be easier to use the same targets for EB rules and EB schedules, right?

> const target = new targets.LambdaInvoke(props.func, {

Author: > Can you explain what this resource achieves that the schedule class in events does not?

In EB Rules a `Schedule` is just a schedule expression – a string that defines on what schedule the rule should be invoked. 

In EB Scheduler a `Schedule` is a separate CFN resource. It has much more properties than just a schedule expression (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-scheduler-schedule.html). For example, `Schedule` supports timezone, groups, encryption, it can be disabled. `Schedule` in Scheduler plays similar role to  `Rule`: it is a central concept of the package that connects event input and event target.

Also, in the prototype implementation of L2 Scheduler constructs, we have a class `ScheduleExpression` that inherits `eb.Schedule`. We have added a method `at(..)` that allows to create a one-time schedule from a date. See https://github.com/filletofish/cdk-eb-scheduler/blob/main/lib/aws-scheduler/lib/expression.ts#L10

    input: ScheduleTargetInput.fromObject({
    "payload": "useful"
    })
});
    
const schedule = new Schedule(this, 'Schedule', {

>     schedule: ScheduleExpression.rate(Duration.minutes(10)),

Author: Yes, see previous comment (https://github.com/aws/aws-cdk-rfcs/pull/486#discussion_r1141427352) - we are inheriting EB `Schedule` class to add implementation of `at(...)` method to create a one-time schedules, which are only supported in Scheduler (AFAIU).

>     schedule: ScheduleExpression.rate(Duration.minutes(10)),

Reviewer: We can still do that, if the `events.Schedule` class was in core. I guess I understand that there is a `Schedule` prop in `events` (and other places) and then a `Schedule` resource that has a `ScheduleExpression` in this RFC, but I think they are the same thing represented in different ways.

A big win for the CDK is that we can take these two different _ways_ of expressing a concept like "schedule" and expose only one CDK "schedule". Then behind the scenes we do the translation into what the different resources want. That is sort of what you are doing in https://github.com/filletofish/cdk-eb-scheduler/blob/main/lib/aws-scheduler/lib/expression.ts#L10. Looks like you aren't creating a brand new `ScheduleExpression` class but extending the existing `Schedule`. That is fine for now, and https://github.com/aws/aws-cdk/issues/8979 can continue to be a ticket asking for `events.Schedule` to be moved to `core`.

>     schedule: ScheduleExpression.rate(Duration.minutes(10)),

Reviewer: No, what you are doing is enough I think.

```ts
class SpecialSchedule extends core.Schedule {
  at(...) {}
}

});
```

### Schedule Expressions

Both rate-based and cron-based schedules are recurring schedules. You configure each recurring schedule type using a schedule expression for the type of schedule you want to configure, and specifying a time zone in which EventBridge Scheduler evaluates the expression.

```ts
const rateBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test rate-based schedule',
});

const cronBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.cron({ day: '20', month: '11' }),
    target,
    scheduleTimeZone: 'America/New_York'
    description: 'This is a test cron-based schedule in New York timezone',
});
```

A one-time schedule is a schedule that invokes a target only once. You configure a one-time schedule when by specifying the time, date, and time zone in which EventBridge Scheduler evaluates the schedule.

```ts
const oneTimeSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.at(new Date(2022, 10, 20, 19, 20, 23)),
    target,

>     scheduleTimeZone: 'America/New_York'

Reviewer: TimeZone enum is added to core: https://github.com/aws/aws-cdk/pull/24127. We should use it.

Author: Amazing, thanks!

    description: 'This is a one-time schedule in New York timezone',
});
```

### Grouping Schedules

When creating a new schedule, you can add it to a scheduling group:

```ts
const group = new Group(this, "Group", {

>     groupName: "MyGroup"

Reviewer: Are there going to be more props to a group? Or is it simply a name.

Also, should we add APIs like `group.addSchedules(schedules: Schedule[])`?

Author: > Are there going to be more props to a group? Or is it simply a name.

By them you mean the owners of AWS EB Scheduler service? I do not know, how can we check that? 

> Also, should we add APIs like group.addSchedules(schedules: Schedule[])?

Sounds like a good idea! 

A question for implementation - currently, we create `CfnSchedule` in a constructor of `Schedule` and pass `groupName` parameter to `CfnSchedule`. How can we delay creation of `CfnSchedule` so that we can change the `groupName` parameter dynamically? Is there a good example of that behaviour somewhere in CDK lib already?

Reviewer: Yes. I think you're asking for Lazy :). We use it all over the place. [here](https://github.com/aws/aws-cdk/blob/cabff71139bf6a3b8ecd4cdf4a8dea093e67b8e6/packages/%40aws-cdk/aws-events/lib/rule.ts#L112) is an example.

});

const target = new targets.LambdaInvoke(props.func, {
    input: ScheduleTargetInput.fromObject({
    "payload": "useful"
    })
});
    
const schedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    group: group,
});
```

If group is not provided schedule will be created in default group, which you can access in CDK with: 
```ts
const defaultGroup = Group.fromDefaultGroup(this, "DefaultGroup");
```

> ## Scheduler Targets

Author: See discussion above - https://github.com/aws/aws-cdk-rfcs/pull/486#discussion_r1141424181

});
```

### Schedule Expressions

Both rate-based and cron-based schedules are recurring schedules. You configure each recurring schedule type using a schedule expression for the type of schedule you want to configure, and specifying a time zone in which EventBridge Scheduler evaluates the expression.

```ts
const rateBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test rate-based schedule',
});

const cronBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.cron({ day: '20', month: '11' }),
    target,
    scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,
    description: 'This is a test cron-based schedule in New York timezone',
});
```

A one-time schedule is a schedule that invokes a target only once. You configure a one-time schedule when by specifying the time, date, and time zone in which EventBridge Scheduler evaluates the schedule.

```ts
const oneTimeSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.at(new Date(2022, 10, 20, 19, 20, 23)),
    target,

>     scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,

Reviewer 2: I need to check. Because I am reusing a class that already has `cron`, I only added `at`.

>     scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,

Author: >  seems a bit odd to have schedule and scheduleTimeZone since scheduleTimeZone seems to be operating on whatever information is in schedule

Yeah, I agree it's odd. However, I don't think we have any other option if we want to reuse class `Schedule` from `events` module, as class `Schedule` already has a method `cron` that does not accept a timezone. Alternatively, we could create our own class `ScheduleExpression` that does not inherit from `Schedule`. What do you think?

>     scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,

Reviewer: Lets just go with as is for now. I'm fine with `scheduleTimeZone`.

    description: 'This is a one-time schedule in New York timezone',
});
```

### Grouping Schedules

> When creating a new schedule, you can add it to a scheduling group:

Reviewer 2: Reasons for grouping:
- access control
- list all schedules in a group
- delete group -> delete all schedules in it

```ts
const group = new Group(this, "Group", {
    groupName: "MyGroup"
});

const target = new targets.LambdaInvoke(props.func, {
    input: ScheduleTargetInput.fromObject({
        "payload": "useful"
    })
});
    
const schedule1 = new Schedule(this, 'Schedule1', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target
});

const schedule2 = new Schedule(this, 'Schedule2', {
    schedule: ScheduleExpression.rate(Duration.minutes(5)),

>     target

Reviewer 2: Yeah we are still playing around in a separate repo. Once the APIs are solid we will move to aws-cdk fork

});

group.addSchedules(schedule1, schedule2);
```

If group is not provided schedule will be created in default group, which you can access in CDK with: 
```ts

> const defaultGroup = Group.fromDefaultGroup(this, "DefaultGroup");

Reviewer 2: Only for being able to do grantRead / grantWrite on the default group :-)

> const defaultGroup = Group.fromDefaultGroup(this, "DefaultGroup");

Author: > We'll have to worry about dependencies because we will use the defaultGroup resource as if it already exists and not as a to-be-created resource.

Yeah, as Jacco mentioned above the default group already exists in every AWS account. 

See https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-schedule-group.html

> Only for being able to do grantRead / grantWrite on the default group :-)

And also for getting its metrics.

> const defaultGroup = Group.fromDefaultGroup(this, "DefaultGroup");

Reviewer: Sounds good. If it already exists, shouldn't be a dependency problem then.

1. `targets.LambdaInvoke`: Invoke an AWS Lambda function
1. `targets.SqsSendMessage`: Send a message to an Amazon SQS Queue
1. `targets.SnsPublish`: Publish a message into an SNS topic

A universal target is a customizable set of parameters that allow you to invoke a wider set of API operation for many AWS services. 
To create a universal target you need to specify `input`, service name and service action. See [full list of support universal targets](https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-targets-universal.html#supported-universal-targets).

For example, to create a schedule that will create a SQS queue every hour:

```ts
const input = ScheduleTargetInput.fromObject({
    'QueueName': 'MyQueue'
});

const schedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.hours(1)),
    target: new targets.Universal('sqs', 'CreateQueue', { input: input })
});
```

### Input 

Target can be invoked with a custom input. Class `ScheduleTargetInput` supports free form text input and JSON-formatted object input:

```ts
const input = ScheduleTargetInput.fromObject({
    'QueueName': 'MyQueue'
});
```

> You can include context attributes in your target payload. EventBridge Scheduler will replace each keyword with its respective value and deliver it to the target. Supported context attributes: 

Reviewer: Can we link to any relevant EB docs here?

4. `ContextAttribute.attemptNumber()` – A counter that identifies the attempt number for the current invocation, for example, 1.

```ts
const text = `Attempt number: ${ContextAttribute.attemptNumber}`;
const input = scheduler.ScheduleTargetInput.fromInput(text);
```

### Specifying IAM role 

An execution role is an IAM role that EventBridge Scheduler assumes in order to interact with other AWS services on your behalf.

The classes for templated schedule targets automatically create an IAM role with all the minimum necessary
permissions to interact with the templated target. If you wish you may specify your own IAM role, then the templated targets 
will grant minimal required permission. For example: for invoking Lambda function target `LambdaInvoke` will grant
execution IAM role permission to `lambda:InvokeFunction`.

```ts
import * as iam from '@aws-cdk/aws-iam';

const role = new iam.Role(this, 'Role', {
  assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
});

const target = new targets.LambdaInvoke({
    input: ScheduleTargetInput.fromObject({ 
        "payload": "useful"
    }),
    role: role

> }, props.func);

Reviewer: replace `props.func` with `declare const fn: lambda.Function;`

    }
```

![L2 Constructs Class Diagram](../images/EventBridge-Scheduler-2023-03-05-1723.excalidraw.png)

Prototype at https://github.com/filletofish/cdk-eb-scheduler/

### Is this a breaking change?

No.

### What alternative solutions did you consider?

> Briefly describe alternative approaches that you considered. If there are
> hairy details, include them in an appendix.

1. We have discussed an opportunity of reusing existing `events.targets` module: https://docs.aws.amazon.com/cdk/api/v1/docs/aws-events-targets-readme.html. See RFC comments

### What are the drawbacks of this solution?

> Describe any problems/risks that can be introduced if we implement this RFC.

TBD

### What is the high-level project plan?

> Describe your plan on how to deliver this feature from prototyping to GA.
> Especially think about how to "bake" it in the open and get constant feedback
> from users before you stabilize the APIs.
>
> If you have a project board with your implementation plan, this is a good
> place to link to it.

Most part of the L2 constructs is already implemented. We have a set of relatively (1-2 day of work) todos that can be implemented in parallel: 

> 1. Decide what removal policy for Group should be 

Reviewer: Do we need to discuss this any further?

Reviewer 2: We need to do some tests. Main point to figure out is if a group you create will actually delete all schedules in it when the group is deleted. That would make it a resource with state and its state would deserve protection.

3. **Schedule Group**: Scheduling groups help you organize your schedules. Scheduling groups support tags that you can use for cost allocation, access control, and resource organization. When creating a new schedule, you can add it to a scheduling group.

This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project. It allows you to define Event Bridge Schedules.

## Defining a schedule 

```ts
const target = new targets.LambdaInvoke(props.func, {
    input: ScheduleTargetInput.fromObject({
        "payload": "useful"
    })
});
    
const schedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test schedule that invokes lambda function every 10 minutes.',
});
```

### Schedule Expressions

> Both rate-based and cron-based schedules are recurring schedules. You configure each recurring schedule type using a schedule expression for the type of schedule you want to configure, and specifying a time zone in which EventBridge Scheduler evaluates the expression.

Author: No, just those two. Will update the doc to make it clear

    target,
    scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,
    description: 'This is a one-time schedule in New York timezone',
});
```

### Grouping Schedules

Your AWS account comes with a default scheduler group. If group is not provided a schedule will be created in default group, which you can access in CDK with: 

```ts
const defaultGroup = Group.fromDefaultGroup(this, "DefaultGroup");
```

When creating a new schedule, you can also add the schedule to a custom scheduling group managed by you:

```ts
const group = new Group(this, "Group", {
    groupName: "MyGroup"
});

const target = new targets.LambdaInvoke(props.func, {
    input: ScheduleTargetInput.fromObject({
        "payload": "useful"
    })
});
    
const schedule1 = new Schedule(this, 'Schedule1', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target
});

const schedule2 = new Schedule(this, 'Schedule2', {
    schedule: ScheduleExpression.rate(Duration.minutes(5)),
    target
});

group.addSchedules(schedule1, schedule2);
```

Groups can be used to organize the schedules logically, access the schedule metrics and manage permissions at group granularity (see below).

## Scheduler Targets

> The `@aws-cdk/aws-schedule-targets` (ADD LINK) module includes classes that implement the `IScheduleTarget` interface for various AWS services.

Author: Wanted to add a link once the module is realised. Better to remove for now. Thanks for highlighting

4. `ContextAttribute.attemptNumber()` – A counter that identifies the attempt number for the current invocation, for example, 1.

```ts
const text = `Attempt number: ${ContextAttribute.attemptNumber}`;
const input = scheduler.ScheduleTargetInput.fromInput(text);
```

### Specifying IAM role 

An execution role is an IAM role that EventBridge Scheduler assumes in order to interact with other AWS services on your behalf.

The classes for templated schedule targets automatically create an IAM role with all the minimum necessary
permissions to interact with the templated target. If you wish you may specify your own IAM role, then the templated targets 
will grant minimal required permission. For example: for invoking Lambda function target `LambdaInvoke` will grant
execution IAM role permission to `lambda:InvokeFunction`.

```ts
import * as iam from '@aws-cdk/aws-iam';

declare const fn: lambda.Function;

const role = new iam.Role(this, 'Role', {
  assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
});

const target = new targets.LambdaInvoke({
    input: ScheduleTargetInput.fromObject({ 
        "payload": "useful"
    }),
    role: role
}, fn);
```

> For universal targets you must grant the required IAM permissions yourself.

Author: Updated the RFC

    target,
    key: key,
});
```

See: [Data Protection](https://docs.aws.amazon.com/scheduler/latest/UserGuide/data-protection.html) in
the *Amazon EventBridge Scheduler User Guide*.

## Error-handling 

You configure how your schedule handles failures, when EventBridge Scheduler is unable to deliver an event
successfully to a target, by using two primary mechanisms: a retry policy, and a dead-letter queue (DLQ). 

A retry policy determines the number of times EventBridge Scheduler must retry a failed event, and how long 
to keep an unprocessed event. 

A DLQ is a standard Amazon SQS queue EventBridge Scheduler uses to deliver failed events to, after the retry 
policy has been exhausted. You can use a DLQ to troubleshoot issues with your schedule or its downstream target. 
If you've configured a retry policy for your schedule, EventBridge Scheduler delivers the dead-letter event after
exhausting the maximum number of retries you set in the retry policy.

```ts
const dlq = new Queue(this, "DLQ", {
      queueName: 'MyDLQ',
});

const target = new targets.LambdaInvoke(props.func, {
    deadLetterQueue: dlq,
    maximumEventAge: Duration.minutes(1),

>     maximumRetryAttempts: 3

Author: In CFN only maximum event age and maximum retry attempts are part of retry policy. Would you recommend to organise them similarly in a `RetryPolicy` structure as well? 

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-retrypolicy.html

1. We have discussed an opportunity of reusing existing `events.targets` module: https://docs.aws.amazon.com/cdk/api/v1/docs/aws-events-targets-readme.html. See RFC comments

### What are the drawbacks of this solution?

> Describe any problems/risks that can be introduced if we implement this RFC.

TBD

### What is the high-level project plan?

> Describe your plan on how to deliver this feature from prototyping to GA.
> Especially think about how to "bake" it in the open and get constant feedback
> from users before you stabilize the APIs.
>
> If you have a project board with your implementation plan, this is a good
> place to link to it.

Most part of the L2 constructs is already implemented. We have a set of relatively (1-2 day of work) todos that can be implemented in parallel: 

1. Decide what removal policy for Group should be 
2. Support and Test Encrypted Schedules
3. Decide or clean up how IAM Roles are used in Schedules and Targets
4. Check why name generation fails if env is not passed to the Stack
5. Unit tests for class schedule
6. Unit tests for class targets 
7. Unit tests for classes: groups, schedule expressions, input 

> 8. Integration tests #15

Author: Accidental typo during copy paste (copied these items from a list of GH issues).

```ts
const target = new targets.LambdaInvoke(props.func, {
    input: ScheduleTargetInput.fromObject({
        "payload": "useful"
    })
});
    
const schedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test schedule that invokes lambda function every 10 minutes.',
});
```

### Schedule Expressions

You can choose from three schedule types when configuring your schedule: rate-based, cron-based, and one-time schedules. 

Both rate-based and cron-based schedules are recurring schedules. You configure each recurring schedule type using a schedule expression for the type of schedule you want to configure, and specifying a time zone in which EventBridge Scheduler evaluates the expression with `scheduleTimeZone` parameter.

```ts
const rateBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test rate-based schedule',
});

const cronBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11' }),
    target,

>     scheduleTimeZone: TimeZone.AMERICA_NEW_YORK,

Reviewer 3: How about renaming the prop `scheduleTimezone` to `timezone` to make it short? I feel it is already clear enough that this timezone property is for the schedule.

Author: I am concerned that the customer might think that timezone is also applied to other properties. For example, `Schedule` also has properties `startDate` and `endDate`, which should be specified in UTC timezone and do not depend on timezone.

So we have these options: 

1 - Do nothing

Keep interface as-is with a separate `scheduleTimeZone` parameter. It's simple and matches CFN structure https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-scheduler-schedule.html#cfn-scheduler-schedule-scheduleexpressiontimezone

2 - Rename `scheduleTimeZone` to `timeZone` 

Not obvious that the property `timeZone` does not apply to `startDate` and `endDate`

3 -  Add timezone parameter to `ScheduleExpression` methods

Example: 
```ts
const cronBasedSchedule = new Schedule(this, 'Schedule', {
    schedule: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11', timezone: TimeZone.AMERICA_NEW_YORK}),
    target,
    description: 'This is a test cron-based schedule that will run at 11:00 PM, on day 20 of the month, only in November in New York timezone',
});
```

In implementation we will create a class `ScheduleExpression` that will inherit from [`events.Schedule`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-events.Schedule.html) and will have methods `cron(..)`, `at(..)` and `rate(..)` with timezone as an additional parameter. 

Not a bad option. Maybe `timezone` will not be that discoverable as in other options. 

4 - Create a separate interface  with two parameters `expression` and `timezone`

Example: 

```ts
const cronBasedSchedule = new Schedule(this, 'Schedule', {
    scheduleExpression: {
          expression: ScheduleExpression.cron({ minute: '0', hour: '23', day: '20', month: '11' }),
          timeZone: TimeZone.AMERICA_NEW_YORK
     },
    target,
    description: 'This is a test cron-based schedule that will run at 11:00 PM, on day 20 of the month, only in November in New York timezone',
});
```

However, I am not sure what name to choose for the new interface as names `Schedule` and `ScheduleExpression` are already taken. Any recommendations? 

---

All in all, I am probably still leaning to the first as it's the simplest option. Curious if you guys also have any opinion on this question @Jacco, @kaizencc  and @comcalvi?

Reviewer 3: That makes sense, thanks! To me, option 3 or 4 look nice and more cohesive, but at the same time option 1 sounds reasonable. I'm fine with whatever decision is made :)

Reviewer: I think I like option 4, with `scheduleExpression: { schedule: Schedule, timeZone: TimeZone }`

Reviewer 2: Few things to consider: 

How would the original `events.Schedule` be changed if timezone support is added for `aws-events`. I think it would best added `CronOptions` in CDK since it has no meaning for a `rate` expressions `timeZone` has no meaning. It would be kept separate though because in cloudFormation is would probably not be part of the expression but a separate property like here.

So maybe we should still inherit but add the timeZone parameter from `cron` and `at`.

Unflat interfaces give less good code completion in some languages (like python).

Another thing to consider: what if `aws-events` at some point adds another type of schedule that is unsupported by `aws-scheduler`. We could add an error raising override, but maybe we'd better not inherited?

Maybe implement like this:

```ts
import { Duration, TimeZone } from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';

/**
 * Schedule for EventBridge Schedule
 *
 * Note that rates cannot be defined in fractions of minutes.
 *
 * @see https://docs.aws.amazon.com/scheduler/latest/UserGuide/schedule-types.html
 */
export abstract class ScheduleExpression {
  /**
   * Construct a one-time schedule from a date.
   *
   * @param date The date and time to use. The millisecond part will be ignored.
   */
  public static at(date: Date, timeZone?: TimeZone): ScheduleExpression {
    try {
      const literal = date.toISOString().split('.')[0];
      return new LiteralScheduleExpression(`at(${literal})`, timeZone)
    } catch(e) {
      if (e instanceof RangeError){
        throw new Error('Invalid date');
      }
      throw e;
    }
  }

  public static cron(options: events.CronOptions, timeZone?: TimeZone): ScheduleExpression {
    return new LiteralScheduleExpression(events.Schedule.cron(options).expressionString, timeZone);
  }

  public static rate(duration: Duration): ScheduleExpression {
    return new LiteralScheduleExpression(events.Schedule.rate(duration).expressionString);
  }

  public static expression(expression: string, timeZone?: TimeZone) {
    return new LiteralScheduleExpression(expression, timeZone);
  }

  /**
   * Retrieve the expression for this schedule
   */
   public abstract readonly expressionString: string;

  /**
   * Retrieve the timeZone for this schedule
   */
   public abstract readonly expressionTimeZone?: TimeZone;
}

class LiteralScheduleExpression extends ScheduleExpression {
  constructor(public readonly expressionString: string, public readonly expressionTimeZone?: TimeZone) {
    super();
  }

  public _bind() {}
}
```

Author: I like that suggestion @Jacco. It's similar to option 3. I was also thinking about the way to reuse functionality of `events.Schedule` with composition, but couldn't come up with a good enough option like yours. 

I will add it to the RFC. Thanks!

Reviewer: Agree on not inheriting for now. The ideal scenario is still to put a generic Schedule into core that can be extended by the 4 resources that have the concept of schedule today. But that's out of scope of this RFC. 

(autoscaling, app-autoscaling, scheduler, events are the 4 i can think of)

    description: 'This is a test cron-based schedule that will run at 11:00 PM, on day 20 of the month, only in November in New York timezone',
});
```

A one-time schedule is a schedule that invokes a target only once. You configure a one-time schedule when by specifying the time, date, and time zone in which EventBridge Scheduler evaluates the schedule.

```ts
const oneTimeSchedule = new Schedule(this, 'Schedule', {

>     schedule: ScheduleExpression.at(new Date(2022, 10, 20, 19, 20, 23)),

Reviewer 3: I guess this `Date` class will be defined somewhere in CDK, as we cannot use Node's built-in Date class here due to jsii limitation.

Reviewer 2: I think I only implemented it this way when I saw another instance where Date was used in the CDK libraries. Will have a look :-)

Reviewer 2: There are many examples of Date being used. Here is an example:

https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-applicationautoscaling/lib/scalable-target.ts

Reviewer 3: TIL about that! Reading [this doc](https://aws.github.io/jsii/specification/2-type-system/), it seems Date is also  supported (a bit misleading though). I'll archive this 👍

### Why should we _not_ do this?

> Is there a way to address this use case with the current product? What are the
> downsides of implementing this feature?

We are not confident that the service API is fully set in stone and implementing an L2 on
top of the current L1 may be setting us up for changes in the future. We are reaching out
to the service team to get their input and plans for the service. 

It’s a large effort (X devs * 1 week) to invest in a module when we have other pressing
projects. However, the bulk of the effort has been spent already since we have fairly
robust prototypes already implemented.

### What is the technical solution (design) of this feature?

```mermaid
classDiagram
    Expression <|-- Schedule : scheduleExpression
    Key <|-- Schedule : key
    ScheduleTargetBase <|-- Schedule : target
    Group <|-- Schedule : group
    Input <|-- ScheduleTargetBase : input
    Targets <.. ScheduleTargetBase : implements

    class Schedule {
        + String name
        + Duration flexibleTimeWindow 
        + Date startDate
        + Date endDate

>         + String scheduleTimeZone 

Reviewer 3: `String` should be replaced to `TimeZone` :)

Author: Thanks, will fix in the next commit

});
    
const schedule = new Schedule(this, 'Schedule', {
    scheduleExpression: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test schedule that invokes lambda function every 10 minutes.',
});
```

### Schedule Expressions

You can choose from three schedule types when configuring your schedule: rate-based, cron-based, and one-time schedules. 

Both rate-based and cron-based schedules are recurring schedules. You can configure each recurring schedule type using a schedule expression. For cron-based schedule you can specify a time zone in which EventBridge Scheduler evaluates the expression. 

```ts
const rateBasedSchedule = new Schedule(this, 'Schedule', {
    scheduleExpression: ScheduleExpression.rate(Duration.minutes(10)),
    target,
    description: 'This is a test rate-based schedule',
});

const cronBasedSchedule = new Schedule(this, 'Schedule', {
    scheduleExpression: ScheduleExpression.cron(
        { minute: '0', hour: '23', day: '20', month: '11' },
        TimeZone.AMERICA_NEW_YORK

>     ),

Author: nice one, will add

    input: ScheduleTargetInput.fromObject({ 
        "payload": "useful"
    }),
    role,
}, fn);
```

> Class for universal target also automatically creates an IAM role if you do no specify your own IAM role. However, in comparison with templated targets, for universal targets you must grant the required IAM permissions yourself.

Reviewer 4: ```suggestion
Universal target automatically create an IAM role if you do not specify your own IAM role. However, in comparison with templated targets, for universal targets you must grant the required IAM permissions yourself.
```
